doc/developer/design/20230209_invalid_accumulations.md

```diff
@@ -0,0 +1,78 @@
+2. Another reason for error due to negative multiplicities is that a negative accumulated result cannot be coerced to an [unsigned type](https://github.com/MaterializeInc/materialize/blob/9933639ea5393d1aaa40c1079d07ffc1cc622516/src/compute/src/render/reduce.rs#L1344-L1346). This error is now treated as unrecoverable, leading to a `panic!` instead of the above strategy of employing a soft assertion.
+3. Following a similar error reporting strategy as in category 1 above, we have a soft assertion regarding net-zero records with [non-zero accumulation in accumulable reductions](https://github.com/MaterializeInc/materialize/blob/9933639ea5393d1aaa40c1079d07ffc1cc622516/src/compute/src/render/reduce.rs#L1283-L1290). This category of error can emerge if we determine that an aggregate row has been removed, due to the corresponding tuple multiplicity accumulating to zero, but there is an inconsistency with the accumulation state, namely the aggregation function on the raw changes ends up with a non-zero result.
+4. As observed in an incident, shutdown of a `persist_source` operator could trigger errors due to the sanity checks above in the COMPUTE layer. This behavior occurs because `persist_source` would not emit the entirety of a batch nor guarantee that a partially emitted batch would consolidate to a multiset. We ignore this problem in the remainder of this document, as a solution was introduced by PR [#17147](https://github.com/MaterializeInc/materialize/pull/17147).
+
+The errors in categories 1 and 3 above are now reported in Sentry and require investigation. Importantly, these errors are concerning as they indicate that Materialize might silently compute incorrect aggregation results. The errors in category 2 cause Materialize to crash in production.
+
+### Approach: Report SQL-level Errors, Log to Sentry, and Eschew Crashes
+
+We discuss below a path for avoiding system crashes and some classes of silently erroneous computation by producing query-level errors whenever invalid accumulation errors are detected in reductions. Given that bugs in other Materialize components, and not only erroneous user-provided source data, could generate invalid accumulations in reductions, we still advocating keeping error reporting to Sentry in addition to producing query-level errors.
```

Comment:
Errors in Sentry could also be useful to the support team, to pro-actively notify users about invalid data in their sources (see [Slack thread](https://materializeinc.slack.com/archives/CM7ATT65S/p1675358436801089)). I would argue this is a more important reason for keeping them in Sentry, since I would assume that developers would mostly ignore these errors when they are known to be cause-able by invalid input -- that is unless we have a good/fast way of ruling out invalid input.

Reply:
Good point, I am adding a sentence about this with the link you provided to the Slack discussion.